Dialogs/PreferencesDialog.cpp

```diff
@@ -1,22 +1,88 @@
 
-  foreach (QString styleName, QStyleFactory::keys()) {
-    ui->styleCombo->addItem(styleName);
-    if (style()->objectName().toLower() == styleName.toLower())
-      ui->styleCombo->setCurrentIndex(ui->styleCombo->count() - 1);
-  }
+  foreach (QString styleName, QStyleFactory::keys()) { ui->styleCombo->addItem(styleName); }
+
+  this->reset();
+
+  connect(ui->buttonBox->button(QDialogButtonBox::RestoreDefaults), &QAbstractButton::clicked, this,
+          &PreferencesDialog::restoreDefaultsClicked);
+  connect(ui->buttonBox->button(QDialogButtonBox::Apply), &QAbstractButton::clicked, this,
+          &PreferencesDialog::applyClicked);
 }
 
 PreferencesDialog::~PreferencesDialog() { delete ui; }
 
-void PreferencesDialog::on_styleCombo_activated(const QString &styleName) {
+static inline QString preferencesKey() { return QStringLiteral("Preferences"); }
+static inline QString generalKey() { return QStringLiteral("General"); }
+static inline QString appearanceKey() { return QStringLiteral("Appearance"); }
+
+void PreferencesDialog::apply() {
+  QSettings settings;
+  settings.beginGroup(preferencesKey());
+
+  settings.beginGroup(generalKey());
+  settings.setValue("documentationURI", ui->documentationLineEdit->text());
+  settings.setValue("websiteURI", ui->websiteLineEdit->text());
+  settings.setValue("communityURI", ui->communityLineEdit->text());
+  settings.setValue("submitIssueURI", ui->submitIssueLineEdit->text());
+  settings.endGroup();  // Preferences/General
+
+  settings.beginGroup(appearanceKey());
+  const QString &styleName = ui->styleCombo->currentText();
+  settings.setValue("styleName", styleName);
   QStyle *style = QStyleFactory::create(styleName);
   if (style) QApplication::setStyle(style);
+  settings.endGroup();  // Preferences/Appearance
+
+  settings.endGroup();  // Preferences
 }
+
+void PreferencesDialog::reset() {
+  QSettings settings;
+  settings.beginGroup(preferencesKey());
+
+  settings.beginGroup(generalKey());
+  ui->documentationLineEdit->setText(
+      settings.value("documentationURI", "https://enigma-dev.org/docs/Wiki/Main_Page").toString());
+  ui->websiteLineEdit->setText(settings.value("websiteURI", "https://enigma-dev.org").toString());
+  ui->communityLineEdit->setText(settings.value("communityURI", "https://enigma-dev.org/forums/").toString());
+  ui->submitIssueLineEdit->setText(
+      settings.value("submitIssueURI", "https://github.com/enigma-dev/RadialGM/issues").toString());
+  settings.endGroup();  // Preferences/General
+
+  settings.endGroup();  // Preferences
+
+  foreach (QString styleName, QStyleFactory::keys()) {
+    if (style()->objectName().toLower() == styleName.toLower()) ui->styleCombo->setCurrentText(styleName);
+  }
+}
+
+void PreferencesDialog::applyClicked() { this->apply(); }
+
+void PreferencesDialog::restoreDefaultsClicked() {
+  QSettings settings;
+  settings.beginGroup(preferencesKey());
+  settings.remove("");

```

Comment:
I'm not familiar with this idiom; what does removing an empty string do? Apparently, it clears the whole group. I'd comment stuff like this. But then, I guess it might not be unreasonable to expect some prior knowledge... so I won't insist.

Reply:
I'll comment it so that the intention is clear. But yeah, since we are "restoring defaults" we just want to remove the whole group.

Reply:
I've shortened and used the key helper so that it becomes clear and obvious: https://github.com/enigma-dev/RadialGM/pull/25/commits/4a490abae0b9e7cd73ed45a145c472ca718d42fd#diff-72e7877bb22fe5bbf3b7eb07dfad31ebR75